{{cookiecutter.project_slug}}/{{cookiecutter.project_slug}}/users/tests/test_views.py

```diff
@@ -50,3 +56,28 @@ def test_get_redirect_url(
         view.request = request
 
         assert view.get_redirect_url() == f"/users/{user.username}/"
+
+
+class TestUserDetailView:
+    def test_user_detail_view(self, rf):
+        user = UserFactory(username="tEsTcAsE")
+        request = rf.get(
+            path=reverse(
+                viewname="users:detail", kwargs={"username": user.username.lower()}
+            )
+        )
+        request.user = user
+        response = UserDetailView.as_view()(request, **{"username": user.username})
+
+        assert response.status_code == 200
+
+    def test_user_detail_view_url(self, client):
+        user = UserFactory(username="tEsTcAsE")
+        response = client.get(
+            path=reverse(
+                viewname="users:detail", kwargs={"username": user.username.lower()}
+            )
+        )
+
+        assert response.status_code == 302
+        assert response.url == f"/accounts/login/?next=/users/{user.username}/"
```

Comment:
I think this change should make the test pass:

```suggestion
        assert response.url == f"/accounts/login/?next=/users/{user.username.lower()}/"
```

I'm not sure there is a bug after all...

Reply:
Sorry for the delay. If the canonical URL for the user is /users/tEsTcAsE, then shouldn't the response URL also be '/users/tEsTcAsE'?

Reply:
>  I don't know how to check the equivalent response URL for the first test case. I just get a 200.

Actually, I realise that this test is probably not testing what we should. It's redirecting because the `UserDetailView` inherits from the `LoginRequiredMixin`, so this test is not even entering the code we should test IMO.

https://github.com/pydanny/cookiecutter-django/blob/2459392a986a193450144aa3fe14228910183f47/%7B%7Bcookiecutter.project_slug%7D%7D/%7B%7Bcookiecutter.project_slug%7D%7D/users/views.py#L11

To correct this and actually test the view, we should log-in the client before calling `client.get()`. The easier way to do that is to use [the `force_login()` method](https://docs.djangoproject.com/en/2.2/topics/testing/tools/#django.test.Client.force_login):

```
client.force_login(user)
```

Then the response status code should either be 200 or 404. 

If it's 200, the next thing to check would be what happens when we have 2 users with the same username but with different cases, e.g. `tEsTcAsE` and `testcase`.

Reply:
Thanks for your guidance on this PR. I created a second user with lowercase ('testcase') and added force_login() but I still got a 200. I also spent some time with pdb tracing through the Django view code.

I looked at the logic of using lower() in the result and I don't think that's the right way to go because we want to test *for* case sensitivity. Finally, I read through https://code.djangoproject.com/ticket/2273 which makes it clear that Django usernames are case sensitive and will remain so.

In the end I no longer think that this is a bug. If you want to keep the test I could modify it to be a vanilla view test becase I think it's still useful as an example. Either way...

Reply:
Thanks for digging all this. I agree, doesn't seem like a bug. I think this could still be good to have some tests for this view. Could you please just make sure the client is logged in for **both tests** (see my latest suggestion). The `response.status_code` should not be 302 IMO...